cyrw/bus_pico_pio.go

```diff
@@ -0,0 +1,79 @@
+//go:build pico && !cy43nopio
+
+package cyrw
+
+import (
+	"encoding/binary"
+	"machine"
+
+	pio "github.com/tinygo-org/pio/rp2-pio"
+	"github.com/tinygo-org/pio/rp2-pio/piolib"
+)
+
+var _busOrder = binary.LittleEndian
+
+type spibus struct {
+	cs  OutputPin
+	spi piolib.SPI3w
+}
+
+func DefaultNew() *Device {
+	// Raspberry Pi Pico W pin definitions for the CY43439.
+	const (
+		// WL_REG_ON = machine.GPIO23
+		// DATA_OUT  = machine.GPIO24
+		// DATA_IN   = machine.GPIO24
+		// IRQ       = machine.GPIO24 // AKA WL_HOST_WAKE
+		// CLK       = machine.GPIO29
+		// CS        = machine.GPIO25
+		WL_REG_ON = machine.GPIO23
+		DATA_OUT  = machine.GPIO24
+		DATA_IN   = DATA_OUT
+		CLK       = machine.GPIO29
+		CS        = machine.GPIO25
+	)
+	WL_REG_ON.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	CS.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	CS.High()
+	sm, err := pio.PIO0.ClaimStateMachine()
+	if err != nil {
+		panic(err.Error())
+	}
+	spi, err := piolib.NewSPI3w(sm, DATA_IN, CLK, 500_000)
+	if err != nil {
+		panic(err.Error())
+	}
+	spi.EnableStatus(true)
+	err = spi.EnableDMA(true)
+	if err != nil {
+		panic(err.Error())
+	}
+	return New(WL_REG_ON.Set, CS.Set, spibus{
+		cs:  CS.Set,
+		spi: *spi,
+	})
+}
+
+func (d *spibus) cmd_read(cmd uint32, buf []uint32) (status uint32, err error) {
+	d.csEnable(true)
+	err = d.spi.CmdRead(cmd, buf)
```

Comment:
Hmm...that's interesting that spi.CmdRead takes a []uint32 buffer rather than a []byte buffer.  Does PIO chunk things in uint32?

Reply:
Yes, PIO works on 32bit registers, we can chunk things in 8 bits but that would mean more overhead per byte. In fact, the non-pio version also chunks in 32 bits. I had designed the whole SPI interface to accomodate PIO in the future!